templates/bucket-replication-lambda.yaml

```diff
@@ -0,0 +1,200 @@
+    Type: String
+    Description: Synapse username used to create files in Synapse.
+  SynapseApikey:
+    Type: String
+    Description: Synapse API key (associated with SynapseUsername) used to authenticate to Synapse.
+  SynapseStorageLocationId:
+    Type: Number
+    Description: >
+      Integer location ID for the bucket containing the files to be added to Synapse. Note a blank value should
+      be provided the when the stack is created.
+      See https://docs.synapse.org/articles/custom_storage_location.html
+  SynapseNotifyUsers:
+    Type: String
+    Description: >
+      Comma-delimited list of numeric Synapse user IDs to be notified when an attempt to sync to Synapse is made
+      by the Lambda function. Example: the `ownerId` returned by a call to the Synapse REST API /userProfile.
+  ReplicationSourceAccountId:
+    Type: Number
+    Description: ID of the AWS account holding the replication source bucket.
+  SynapseAWSAccountId:
+    Type: Number
+    Description: >
+      ID of the Synapse AWS account. Used for allowing Synapse to access the DataBucket as an external storage location.
+      See https://docs.synapse.org/articles/custom_storage_location.html
+    Default: 325565585839
+  Md5BlockSize:
+    Type: Number
+    Default: 5242880 # 5 MB
+    Description:
+      Maximum block size allocated to computing MD5 sums, in bytes.
+
+Resources:
+
+  LambdaExecutionRole:
+    Type: AWS::IAM::Role
+    Properties:
+      #Description: ARN of role that lambda uses for execution.
+      RoleName: !Sub ${AWS::StackName}-LambdaExecutionRole
+      AssumeRolePolicyDocument:
+        Statement:
+          - Action: ['sts:AssumeRole']
+            Effect: Allow
+            Principal:
+              Service: [lambda.amazonaws.com]
+
+  LambdaExecutionRolePolicy:
+    Type: AWS::IAM::Policy
+    Properties:
+      #Description: Allows access to the S3 data bucket and CloudWatch logs.
+      PolicyName: !Sub ${AWS::StackName}-LambdaExecutionPolicy
+      Roles: [!Ref 'LambdaExecutionRole']
+      PolicyDocument:
+        Statement:
+          - Effect: Allow
+            Action:
+              - logs:CreateLogGroup
+              - logs:CreateLogStream
+              - logs:PutLogEvents
+            Resource: "*"
+          - Effect: Allow
+            Action:
+              - s3:ListBucket
+            Resource:
+              - !Join
+                - ""
+                - - "arn:aws:s3:::"
+                  - !Sub ${AWS::StackName}-data
+            Sid: VisualEditor1
+          - Effect: Allow
+            Action:
+              - s3:PutObject
+              - s3:GetObject
+            Resource:
+              - !Join
+                - ""
+                - - "arn:aws:s3:::"
+                  - !Sub ${AWS::StackName}-data
+                  - "/*"
+
+  BucketToSynapseFunction:
+    Type: AWS::Lambda::Function
+    Properties:
+      FunctionName: !Ref FunctionName
+      Code:
+        S3Bucket: !Ref FunctionCodeBucket
```

Comment:
where is this lambda and how is it being built and deployed?

Reply:
The code is here: https://github.com/ecmonsen/Synapse-replication-uploader

Private repo, just invited you and @tthyer

Reply:
And, I have been building it with [aws sam](https://aws.amazon.com/serverless/sam/) (but not using sam deployment). I have a bash script `update-stack.sh` in that repo that builds, copies to S3, and uses CF to create the stack. Obviously that process will need to be slightly different in sandbox where CF is automated...

Reply:
the code is fairly long: https://github.com/ecmonsen/Synapse-replication-uploader/blob/master/lambda_function.py

And it depends on multiple Python packages, so it cannot be included inline in the template.

I don't know if there is a repo in Sage-Bionetworks for lambda projects, will look into it.